src/gda/ionic/queue_pair_ionic.cpp

```diff
@@ -0,0 +1,312 @@
+/******************************************************************************
+ * Copyright (c) Advanced Micro Devices, Inc. All rights reserved.
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ *****************************************************************************/
+
+#include "gda/queue_pair.hpp"
+#include "gda/endian.hpp"
+#include "util.hpp"
+#include "containers/free_list_impl.hpp"
+
+namespace rocshmem {
+
+__device__ uint64_t QueuePair::get_same_qp_lane_mask() {
+  uint64_t lane_mask = get_active_lane_mask();
+  uintptr_t this_val = reinterpret_cast<uintptr_t>(this);
+
+  // exclude threads operating on a different qp from this thread lane mask
+  #pragma unroll
+  for (int i = 0; i < 64; ++i) {
+    uint64_t bit_i = 1ull << i;
+    if ((lane_mask & bit_i) && __shfl(this_val, i) != this_val) {
+      lane_mask &= ~bit_i;
+    }
+  }
+
+  return lane_mask;
```

Comment:
```suggestion
  // Bitmask of lanes in this warp whose value == this_val
  uint64_t same_qp_mask = __match_any_sync(active, this_val);

  return same_qp_mask;
```

This for loop gets executed every time an RMA or atomic call is made. The `__match_any_sync` warp-level intrinsic achieves the same result as that loop

Reply:
I do not want to introduce actual code change in a PR that rename/shuffles code around, but that is a good point for a followup.